torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -456,10 +466,22 @@ void ProcessGroupAgent::enqueueRecv(RecvWork work) {
 }
 
 void ProcessGroupAgent::listenLoop() {
-  while (true) {
+  while (rpcRunning_.load()) {
     // rank, tensor size, message type
     std::vector<torch::Tensor> preamble = {torch::empty({3}, {torch::kInt64})};
-    pg_->recvAnysource(preamble, pg_->getRank())->wait();
+    {
+      std::lock_guard<std::mutex> guard(recvWorkMutex_);
+      recvWork_ = pg_->recvAnysource(preamble, pg_->getRank());
+    }
+
+    if (!rpcRunning_.load()) {
+      return;
+    }
+    bool aborted = !recvWork_->wait();
```

Comment:
Question: sorry I didn't closely follow the abort work. Does the order of `recvWork_->abort();` and `recvWork_->wait()` matter?

Reply:
In this case, the order should be such as:
T1: Calls into `recvWork->wait()`
T2: Aborts the above with `recvWork->abort()`
T1: returns out of recvWork, finishes thread

Reply:
What if `abort()` is called before T1 calls `wait()`? Will the `wait()` immediately returns with `aborted = true`?

Reply:
Yep, this code eventually calls this in gloo: https://github.com/facebookincubator/gloo/blob/master/gloo/transport/tcp/unbound_buffer.cc#L61-L63

that block would translate to something like
`while(!abortWaitRecv_ && recvCompletions <= 0) {
    wait(lock, timeout);
}`

so if `abort()` is called before T1 calls `wait`, then when T1 calls into wait it will return immediately.